File: C# part 2/StringsAndTextProcessing/03.CorrectBrackets/CorrectBrackets.cs
/*
 Problem 3. Correct brackets

    Write a program to check if in a given expression the brackets are put correctly.

Example of correct expression: ((a+b)/5-d). Example of incorrect expression: )(a+b)).
 */
namespace _03.CorrectBrackets
{
    using System;
    class CorrectBrackets
    {
        static void Main()
        {
            string example = "((a + b) / 5 - d)";
            Console.WriteLine(isCorectBrackets(example));
            example = ")(a+b))";
            Console.WriteLine(isCorectBrackets(example));
        }

        private static bool isCorectBrackets(string exp)
        {
            int countOpenBrackets = 0;
            int countCloseBrackets = 0;
            bool validBrackets = false;
            if ((exp[0] == ')') || (exp[exp.Length - 1] == '('))
            {
                validBrackets = false;
                return validBrackets;
            }
            else
            {
                for (int i = 0; i < exp.Length; i++)
                {
                    if (exp[i] == '(')
                    {
                        countOpenBrackets++;
                    }
                    if (exp[i] == ')')
                    {
                        countCloseBrackets++;
                    }
                }
            }

            if (countCloseBrackets == countOpenBrackets)
            {
                return true;
            }
            return validBrackets;
        }
    }
}

File: JavaScript Fundamentals/Javascript Strings/scripts/correctBrackets.js
/// <reference path="js-console.js"/>

print('Problem 2: Write a JavaScript function to check if in a given expression the brackets are put correctly.');

var expression1 = '((a + b)/5 - d)';
var expression2 = ')(a + b))';

function isExpressionBracketsCorrectly(expression) {
    var i, openBrackets = [];
    var isValid = false;
    for (i = 0; i < expression.length; i++) {
        if (expression[i] === '(') {
            openBrackets.push(expression[i]);
        }
        else if (expression[i] === ')') {
            if (openBrackets.length === 0) {
                return isValid;
            }
            openBrackets.pop();
        }
    }

    if (openBrackets.length === 0) {
        isValid = true;
    }

    return isValid;
}

jsConsole.writeLine(expression1 + ' --> ' + isExpressionBracketsCorrectly(expression1));
jsConsole.writeLine(expression2 + ' --> ' + isExpressionBracketsCorrectly(expression2));
console.log(expression1 + ' --> ' + isExpressionBracketsCorrectly(expression1));
console.log(expression2 + ' --> ' + isExpressionBracketsCorrectly(expression2));